nf_requirements/v1alpha1/nf_topology_types.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2023 The Nephio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	//apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// NFType defines the type of network functions
+type NFType string
+
+const (
+	NFTypeUPF NFType = "upf"
+	NFTypeSMF NFType = "smf"
+	NFTypeAMF NFType = "amf"
+)
+
+// NFAttachment defines the specification of network attachment points of a NF
+type NFAttachment struct {
+	// Name of the network attachment point
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// NetworkInstanceRef is a reference to NetworkInstance. Two NF with attachment to
+	// the same NetworkInstance is considered connected neighbors
+	NetworkInstanceName string `json:"networkInstanceName,omitempty" yaml:"networkInstanceName,omitempty"`
+}
+
+// NFTemplate defines the template for deployment of an instance of a NF
+type NFTemplate struct {
+	// NFType specifies the type of NF this template is specifying
+	NFType NFType `json:"nfType,omitempty" yaml:"nfType,omitempty"`
+
+	// ClassName --- for now, the NFClass this NF template will derive from
+	ClassName string `json:"className,omitempty" yaml:"className,omitempty"`
+
+	// Capacity specifies the NF capacity profile for this NF instance
+	Capacity CapacitySpec `json:"capacity,omitempty" yaml:"capacity,omitempty"`
+
+	// NFAttachments
+	NFAttachments []NFAttachment `json:"nfAttachments,omitempty" yaml:"nfAttachments,omitempty"`
+}
+
+type NFInstance struct {
+	// Name specifies the name of this NFInstance
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// ClusterSelector specifies the matching labels for the NF instance to be instantiated
+	ClusterSelector metav1.LabelSelector `json:"clusterSelector,omitempty" yaml:"clusterSelector,omitempty"`
```

Comment:
if this is optional it should be a pointer

Reply:
This is required, not optional. One of the things I did for simplifying implementation is that the NFInstance is ONLY key'ed off of cluster selector (instead of other potential selector also). I think it should suffice for R1, and can evolve the CRD as we discover other use cases to instantiate NF instances dynamically.

Reply:
understood. We should keep the label selected generic but only use it for cluster-name. We did the same in various functions.